File: tools/parity/dispatch/fleet-exceptions.ts
// TS dispatch for the fleet-exceptions parity module, against the Rust port
// (orca_core::fleet_exceptions).
//
// The renderer twin was CUT OVER: src/renderer/src/components/alab/fleet-exceptions.ts
// keeps the row types, EXCEPTION_SEVERITY and EXCEPTION_SOURCE_STATUS as data,
// and the reducer now lives behind src/renderer/src/lib/git-wasm/fleet-exception-queue.ts.
// This adapter drives that SHIM rather than the wasm oracle, exactly as the
// wsl-paths adapter does: config/vitest.parity.config.ts never marks the core
// ready, so the shim answers from its `parity` fallback — which is the deleted
// body, and the code the renderer runs whenever the wasm has failed. The harness
// therefore keeps a real TS-vs-Rust differential instead of degenerating to
// wasm-vs-binary, and the fallback is the half that most needs one.
//
// `exceptionSourceStatuses` still reads the TS table directly, because the table
// is the data the twin kept: that case is what fails when the TS and Rust copies
// drift, which is the one drift a cut-over could otherwise hide.
//
// The decision under test is the COLLAPSE KEY. An earlier bug keyed the collapse
// on the run instead of the task, so a run with twelve stuck tasks rendered as
// one row and eleven blocked tasks were invisible to the supervisor. The corpus
// carries several exceptions from one run across different tasks precisely so a
// re-introduction of that key shows up on both sides at once.
//
// Both modules are pure and import nothing from React, so they are imported
// directly — no component harness, no DOM.

import {
  collapseExceptionsByTask,
  unwiredExceptionSources
} from '../../../src/renderer/src/lib/git-wasm/fleet-exception-queue'
import {
  EXCEPTION_SOURCE_STATUS,
  type FleetException,
  type FleetExceptionKind
} from '../../../src/renderer/src/components/alab/fleet-exceptions'

const KINDS: readonly string[] = [
  'gate',
  'escalation',
  'circuit-broken',
  'lifecycle-rejected',
  'attention',
  'unanswered-ask'
]

/** A wrong-typed field reads as absent. `String(x ?? '')` coerced `41` to `'41'`
 *  while the Rust adapter's `as_str()` gave `''`, so the two sides were compared
 *  on different task ids. */
function stringField(raw: Record<string, unknown>, key: string): string {
  const value = raw[key]
  return typeof value === 'string' ? value : ''
}

/**
 * `null` for a kind outside §8.3's six. TS would index SEVERITY with it and get
 * `undefined` — every comparison false, the sort comparator NaN — which is not a
 * behaviour either side should be asserted to reproduce, so both adapters refuse
 * the row instead.
 */
function toException(raw: Record<string, unknown>): FleetException | null {
  const kind = typeof raw.kind === 'string' ? raw.kind : ''
  if (!KINDS.includes(kind)) {
    return null
  }
  return {
    taskId: stringField(raw, 'taskId'),
    kind: kind as FleetExceptionKind,
    summary: stringField(raw, 'summary'),
    // Absent and explicit null are the same "no worker", matching the Rust
    // adapter's `Option<String>`.
    workerHandle: typeof raw.workerHandle === 'string' ? raw.workerHandle : null,
    // Safe integers only: a fraction, or a magnitude past 2**53, is a number the
    // two sides cannot agree they were even handed (Rust reads i64, TS reads
    // f64), so both adapters read it as absent rather than assert a false match.
    attempts: Number.isSafeInteger(raw.attempts) ? (raw.attempts as number) : 0,
    at: stringField(raw, 'at')
  }
}

export function dispatch(fn: string, input: unknown): unknown {
  const args = (input ?? {}) as Record<string, unknown>
  switch (fn) {
    case 'collapseExceptionsByTask': {
      const rows = Array.isArray(args.exceptions) ? args.exceptions : []
      const parsed = rows.map((row) => toException((row ?? {}) as Record<string, unknown>))
      if (parsed.some((row) => row === null)) {
        return { __parity_error__: 'unknown FleetExceptionKind' }
      }
      return collapseExceptionsByTask(parsed as FleetException[])
    }
    case 'unwiredExceptionSources':
      return unwiredExceptionSources()
    // Ordered projection of the status object: pins the six keys AND their
    // declaration order, which is the order Object.keys — and therefore
    // unwiredExceptionSources — reports in.
    case 'exceptionSourceStatuses':
      return Object.entries(EXCEPTION_SOURCE_STATUS).map(([kind, status]) => ({ kind, status }))
    default:
      return { __parity_error__: `unknown function ${fn}` }
  }
}

File: rust/crates/orca-dispatch/src/modules/fleet_exceptions.rs
//! Parity dispatch for `orca_core::fleet_exceptions` vs its TS twin
//! `src/renderer/src/components/alab/fleet-exceptions.ts`.
//!
//! The exceptions queue is what a supervisor actually looks at, so the thing
//! this seam protects is the collapse KEY: an earlier bug collapsed per run
//! instead of per task and showed one row for a run with twelve stuck tasks.
//! Both sides run the same corpus, so a key or precedence drift cannot land
//! quietly on one of them.

use orca_core::fleet_exceptions::{
    collapse_exceptions_by_task, exception_source_status, unwired_exception_sources,
    FleetException, FleetExceptionKind, EXCEPTION_SOURCE_ORDER,
};
use serde_json::{json, Value};

fn string_at(input: &Value, key: &str) -> String {
    input
        .get(key)
        .and_then(Value::as_str)
        .unwrap_or_default()
        .to_string()
}

/// Absent or JSON `null` both read as "no worker", matching the TS
/// `workerHandle: string | null`.
fn optional_string_at(input: &Value, key: &str) -> Option<String> {
    input
        .get(key)
        .and_then(Value::as_str)
        .map(ToString::to_string)
}

/// Safe-integer only: past 2^53 the TS half holds an f64 that is no longer the
/// number the vector wrote, so an agreement there would be an artefact of the
/// harness. Both adapters read such a value as absent instead.
fn attempts_at(input: &Value) -> i64 {
    const MAX_SAFE_INTEGER: u64 = 9_007_199_254_740_991;
    input
        .get("attempts")
        .and_then(Value::as_i64)
        .filter(|attempts| attempts.unsigned_abs() <= MAX_SAFE_INTEGER)
        .unwrap_or(0)
}

/// `None` when `kind` is outside §8.3's six. Neither side guesses a severity for
/// an unknown badge; the TS adapter refuses the same input the same way.
fn parse_exception(value: &Value) -> Option<FleetException> {
    Some(FleetException {
        task_id: string_at(value, "taskId"),
        kind: FleetExceptionKind::parse(value.get("kind").and_then(Value::as_str)?)?,
        summary: string_at(value, "summary"),
        worker_handle: optional_string_at(value, "workerHandle"),
        attempts: attempts_at(value),
        at: string_at(value, "at"),
    })
}

fn exception_to_json(row: &FleetException) -> Value {
    json!({
        "taskId": row.task_id,
        "kind": row.kind.as_str(),
        "summary": row.summary,
        "workerHandle": row.worker_handle,
        "attempts": row.attempts,
        "at": row.at,
    })
}

pub fn dispatch(function: &str, input: &Value) -> Value {
    match function {
        "collapseExceptionsByTask" => {
            let raw = input
                .get("exceptions")
                .and_then(Value::as_array)
                .map(Vec::as_slice)
                .unwrap_or_default();
            let parsed: Option<Vec<FleetException>> = raw.iter().map(parse_exception).collect();
            match parsed {
                Some(rows) => Value::Array(
                    collapse_exceptions_by_task(rows)
                        .iter()
                        .map(exception_to_json)
                        .collect(),
                ),
                None => json!({ "__parity_error__": "unknown FleetExceptionKind" }),
            }
        }
        "unwiredExceptionSources" => Value::Array(
            unwired_exception_sources()
                .into_iter()
                .map(|kind| Value::String(kind.as_str().to_string()))
                .collect(),
        ),
        // Projection of the TS `EXCEPTION_SOURCE_STATUS` object to an ordered
        // list, so the corpus pins both the six keys and their declaration order
        // (the order `Object.keys` — and therefore `unwiredExceptionSources` —
        // would report a regression in).
        "exceptionSourceStatuses" => Value::Array(
            EXCEPTION_SOURCE_ORDER
                .into_iter()
                .map(|kind| {
                    json!({ "kind": kind.as_str(), "status": exception_source_status(kind).as_str() })
                })
                .collect(),
        ),
        other => json!({ "__parity_error__": format!("unknown function {other}") }),
    }
}
